gcovr/timestamps.py

```diff
@@ -0,0 +1,206 @@
+
+    Examples of parsing RFC 3339 like timestamps:
+
+    >>> parse_timestamp("2021-12-27 13:05:27").isoformat()  # no prefix
+    '2021-12-27T13:05:27'
+    >>> parse_timestamp("rfc3339:2021-12-27 13:05:27").isoformat()
+    '2021-12-27T13:05:27'
+
+    Examples of invalid formats:
+
+    >>> parse_timestamp("illegal-scheme:foo")
+    Traceback (most recent call last):
+      ...
+    ValueError: unknown timestamp format
+    >>> parse_timestamp("tomorrow at 2 PM")
+    Traceback (most recent call last):
+      ...
+    ValueError: unknown timestamp format
+    """
+
+    if value.startswith("@"):
+        return _parse_epoch(value[1:])
+
+    # handle explicit schemes
+    scheme_match = re.fullmatch(r"([a-zA-Z][a-zA-Z0-9+.-]*):(.*)", value)
+    if scheme_match is not None:
+        scheme = scheme_match.group(1).lower()
+        value = scheme_match.group(2)
+        if scheme == "epoch":
+            return _parse_epoch(value)
+        if scheme == "rfc3339":
+            return _parse_rfc3339(value)
+        raise ValueError("unknown timestamp format")
+
+    # guess the format
+    for parser in [_parse_epoch, _parse_rfc3339]:
+        try:
+            return parser(value)
+        except ValueError:
+            pass
+
+    raise ValueError("unknown timestamp format")
+
+
+def _parse_epoch(value: str) -> datetime.datetime:
+    r"""
+    Parse an epoch timestamp.
+
+    Epoch timestamps are always resolved in the UTC timezone.
+
+    >>> _parse_epoch("1640606727").isoformat()
+    '2021-12-27T12:05:27+00:00'
+    >>> _parse_epoch("invalid")
+    Traceback (most recent call last):
+      ...
+    ValueError: not a valid Unix epoch
+    """
+    try:
+        return datetime.datetime.fromtimestamp(int(value), UTC)
+    except Exception:
+        raise ValueError("not a valid Unix epoch") from None
+
+
+def _parse_rfc3339(value: str) -> datetime.datetime:
+    r"""
+    Parse an RFC-3339 or ISO-like timestamp.
+
+    This will accept timestamps in the following formats:
+
+    * `YYYY-MMM-DD hh:mm:ss`
+    * same, but with `T` separator instead of space
+    * same, but with timezone offset (e.g. `+hh:mm` or `Z`)
+
+    Note that timestamps without an explicit timezone are naive
+    and represent a local time,
+    which may complicate some aspects of testing.
+
+    Examples for separators:
+
+    >>> _parse_rfc3339("2021-12-27 13:05:27").isoformat()
+    '2021-12-27T13:05:27'
+    >>> _parse_rfc3339("2021-12-27T13:05:27").isoformat()
+    '2021-12-27T13:05:27'
+    >>> _parse_rfc3339("2021-12-27@13:05:27+12:30")
+    Traceback (most recent call last):
+      ...
+    ValueError: timestamp separator must be 'T' or space
+
+    Examples for timezone offsets:
+
+    >>> _parse_rfc3339("2021-12-27 13:05:27Z").isoformat()
+    '2021-12-27T13:05:27+00:00'
+    >>> _parse_rfc3339("2021-12-27 13:05:27+12:30").isoformat()
+    '2021-12-27T13:05:27+12:30'
+    >>> _parse_rfc3339("2021-12-27T13:05:27-07:23").isoformat()
```

Comment:
```suggestion
    >>> _parse_rfc3339("2021-12-27t13:05:27-07:23").isoformat()
```
To test also the lowercase separator.

Reply:
Additional test cases were added.